lms/models/h_user.py

```diff
@@ -33,6 +34,26 @@ class HUser(NamedTuple):
     provider_unique_id: str = ""
     """The "provider_unique_id" string to pass to the h API for this user."""
 
+    @classmethod
+    def from_lti_user(cls, lti_user, authority):
+        provider = lti_user.tool_consumer_instance_guid
+        provider_unique_id = lti_user.user_id
+
+        def username():
+            """Return the h username for the current request."""
+            username_hash_object = hashlib.sha1()
+            username_hash_object.update(provider.encode())
+            username_hash_object.update(provider_unique_id.encode())
+            return username_hash_object.hexdigest()[:30]
```

Comment:
I think this would be better as a cached property like `userid`. Then this function would be:
```python
    @classmethod
    def from_lti_user(cls, lti_user, authority):
        return cls(
            authority=authority,
            display_name=lti_user.display_name,
            provider=lti_user.tool_consumer_instance_guid,
            provider_unique_id=lti_user.user_id,
        )
```

Reply:
As per later in this PR, if the `lti_user` had `authority` this could just be an `h_user` property on the `lti_user` as well

Reply:
Unfortunately there are five places in the code where `HUser`'s are created, and the `username` only comes from `provider` and `provider_unique_id` in one of those places, so it can't be a computed property:

1. `LTIHService` is the one place where `from_lti_user()` is called and `username` is computed from `provider` and `provider_unique_id`:

   https://github.com/hypothesis/lms/blob/a9cd1fd77aa8f046bcd8e6989a06a3f7876bf76e/lms/services/lti_h.py#L35-L38

2. `JSConfig` creates `HUser`'s whose `username` comes from the `lis_result_sourcedid` (`GradingInfo`) table in the DB, and who do not have a `provider` or `provider_unique_id`:

   https://github.com/hypothesis/lms/blob/a9cd1fd77aa8f046bcd8e6989a06a3f7876bf76e/lms/resources/_js_config.py#L315-L319

3. `HAPIService` creates the `"lms"` machine user with a hardcoded `username`:

   https://github.com/hypothesis/lms/blob/a9cd1fd77aa8f046bcd8e6989a06a3f7876bf76e/lms/services/h_api.py#L118-L120

4. `HAPIService` creates an `HUser` with a given `username` and no `provider` or `provider_unique_id`, in order to generate a `userid` from it. This is related to the non-Canvas grading feature again:

   https://github.com/hypothesis/lms/blob/a9cd1fd77aa8f046bcd8e6989a06a3f7876bf76e/lms/services/h_api.py#L36

5. `HAPIService` creates an `HUser` with a `username` and `display_name` received from the h API and no `provider` or `provider_unique_id`. Non-Canvas grading, again:

   https://github.com/hypothesis/lms/blob/a9cd1fd77aa8f046bcd8e6989a06a3f7876bf76e/lms/services/h_api.py#L39-L43

I think the factory method approach is more flexible and, given how `HUser` is used, is appropriate:

1. If you have an `LTIUser` and want an `HUser` from it you use `HUser.from_lti_user()` and `username` is computed for you from `provider` and `provider_unique_id`

2. If you want to create an `HUser` with a given `username` (and potentially without a `provider` or `provider_unique_id`) you do `HUser(username="...", ...)`

Reply:
Fair enough, it seems it's more often _not_ created this way than created this way. This gives me a few thoughts:

I can't see any reason not to do what we did on the previous PR here. So remove the factory method, have a free floating `username_from_provider`. This would also mean we don't have the "chase the code" in LTI-H. If LTI-H is going to be responsible for this it should probably do the work. It would end up pretty short, and eliminate another interface causing testing woes. Maybe something like:

```
@property
def h_user(self):
    """Return the h user for the current request's LTI user."""
    provider = self._lti_user.tool_consumer_instance_guid,
    provider_unique_id = self._lti_user.user_id,

    return HUser(
        authority=self._authority,
        username=username_from_provider(provider, provider_unique_id),
        display_name=self._lti_user.display_name,
        provider=provider,
        provider_unique_id=provider_unique_id,
    )
```

I think if we hadn't started with a `NamedTuple` we wouldn't be having the trouble we are. A normal class could accomodate all of the access patterns above, but creating username when it's null. Maybe something like:

```
class HUser:
    def __init__(self, authority, display_name, provider, provider_unique_id, username=None):
        self.authority = authority
        self.display_name = display_name
        self.provider = provider
        self.provider_unique_id = provider_unique_id

        self.username = self._make_username(provider, provider_unique_id) \
            if username is None else username

    @staticmethod
    def _make_username(provider, provider_unique_id):
        """Return the h username for the current request."""
        digest = hashlib.sha1()
        digest.update(provider.encode())
        digest.update(provider_unique_id.encode())

        return digest.hexdigest()[:30]
```